cpp/include/utilities/comm_utils.cuh

```diff
@@ -0,0 +1,788 @@
+  {
+    using element_t = typename thrust::tuple_element<I, TupleType>::type;
+    static_assert(sizeof(element_t) <= sizeof(int64_t));
+    auto ptr              = reinterpret_cast<element_t const*>(tuple_scalar_elements.data() + I);
+    thrust::get<I>(tuple) = *ptr;
+    update_tuple_from_vector_of_tuple_scalar_elements_impl<TupleType, I + 1, N>().update(
+      tuple, tuple_scalar_elements);
+  }
+};
+
+template <typename TupleType, size_t I>
+struct update_tuple_from_vector_of_tuple_scalar_elements_impl<TupleType, I, I> {
+  void update(TupleType& tuple, std::vector<int64_t> const& tuple_scalar_elements) const { return; }
+};
+
+template <typename TupleType, size_t I, size_t N>
+struct host_allreduce_tuple_scalar_element_impl {
+  void run(raft::comms::comms_t const& comm,
+           rmm::device_uvector<int64_t>& tuple_scalar_elements,
+           cudaStream_t stream) const
+  {
+    using element_t = typename thrust::tuple_element<I, TupleType>::type;
+    static_assert(sizeof(element_t) <= sizeof(int64_t));
+    auto ptr = reinterpret_cast<element_t*>(tuple_scalar_elements.data() + I);
+    comm.allreduce(ptr, ptr, 1, raft::comms::op_t::SUM, stream);
+    host_allreduce_tuple_scalar_element_impl<TupleType, I + 1, N>().run(
+      comm, tuple_scalar_elements, stream);
+  }
+};
+
+template <typename TupleType, size_t I>
+struct host_allreduce_tuple_scalar_element_impl<TupleType, I, I> {
+  void run(raft::comms::comms_t const& comm,
+           rmm::device_uvector<int64_t>& tuple_scalar_elements,
+           cudaStream_t stream) const
+  {
+  }
+};
+
+template <typename T>
+T* iter_to_raw_ptr(T* ptr)
+{
+  return ptr;
+}
+
+template <typename T>
+T* iter_to_raw_ptr(thrust::device_ptr<T> ptr)
+{
+  return thrust::raw_pointer_cast(ptr);
+}
+
+template <typename T>
+auto iter_to_raw_ptr(thrust::detail::normal_iterator<thrust::device_ptr<T>> iter)
+{
+  return thrust::raw_pointer_cast(iter.base());
+}
+
+template <typename InputIterator, typename OutputValueType>
+std::enable_if_t<std::is_same<OutputValueType, thrust::detail::any_assign>::value, void>
+device_isend_impl(raft::comms::comms_t const& comm,
+                  InputIterator input_first,
+                  size_t count,
+                  int dst,
+                  int tag,
+                  raft::comms::request_t* request)
+{
+  // no-op
+}
+
+template <typename InputIterator, typename OutputValueType>
+std::enable_if_t<std::is_arithmetic<OutputValueType>::value, void> device_isend_impl(
+  raft::comms::comms_t const& comm,
+  InputIterator input_first,
+  size_t count,
+  int dst,
+  int tag,
+  raft::comms::request_t* request)
+{
+  static_assert(
+    std::is_same<typename std::iterator_traits<InputIterator>::value_type, OutputValueType>::value);
+  comm.isend(iter_to_raw_ptr(input_first), count, dst, tag, request);
+}
+
+template <typename InputIterator, typename OutputIterator, size_t I, size_t N>
+struct device_isend_tuple_iterator_element_impl {
```

Comment:
What's the reason for having this file in cuGraph over RAFT?

Reply:
Time :-(

The best is NCCL building a C++ extension based on thrust iterator, it will be the best. Yes, second best is move this to RAFT when we move the pattern accelerator API to raft. For this release, we may keep this here to save time.